inference-engine/src/mkldnn_plugin/mkldnn_graph.cpp

```diff
@@ -1197,21 +1197,67 @@ bool MKLDNNGraph::InsertNode(MKLDNNNodePtr parent, MKLDNNNodePtr child, MKLDNNNo
 void MKLDNNGraph::EnforceBF16() {
     // Floating point parts of FP32 + INT8 or FP32 + BIN mixed precision models will be executed in BF16 precision
     // only if enforceBF16 flag was set manually because current performance is not good enough to enable it by default
-    if (implication(isQuantized(), config.manualEnforceBF16)) {
-        for (auto &node : graphNodes) {
-            if (node->getType() != Input && node->getType() != Output) {
-                for (size_t i = 0; i < node->getOriginalInputsNumber(); i++) {
-                    auto &parent = node->getParentEdgesAtPort(i)[0]->getParent();
-                    if (!(parent->getType() == Input && parent->isConstant()) &&       // exclude nodes after Constant Inputs
-                        !(parent->getType() == Input && node->getType() == Eltwise) && // exclude Eltwise after Input since it supports conversion to BF16
-                        node->getOriginalInputPrecisionAtPort(i) == Precision::FP32)
-                        node->setOriginalInputPrecisionAtPort(i, Precision::BF16);
-                }
+    if (!implication(isQuantized(), config.manualEnforceBF16))
+        return;
+    /* list of node types that must be forced to be executed in BF16 precision
+     * because of performance gains */
+    static const std::unordered_set<Type> significantNodes {
+        Convolution,
+        FullyConnected,
+        RNNCell,
+        RNNSeq,
+        MatMul,
+        ROIPooling,
+    };
 
-                for (size_t i = 0; i < node->getOriginalOutputsNumber(); i++) {
-                    if (node->getOriginalOutputPrecisionAtPort(i) == Precision::FP32)
-                        node->setOriginalOutputPrecisionAtPort(i, Precision::BF16);
-                }
+    auto shouldNotBeSkipped = [&](const MKLDNNNodePtr& node) {
+        return significantNodes.count(node->getType());
+    };
+
+    std::function<void(const MKLDNNNodePtr&, std::unordered_set<MKLDNNNodePtr>& skipNodes)> searchForNodesToSkip;
+    searchForNodesToSkip = [&](const MKLDNNNodePtr& node, std::unordered_set<MKLDNNNodePtr>& skipNodes) -> void {
+        for (const auto& edge : node->getParentEdges()) {
+            const auto& parent = edge.lock()->getParent();
```

Comment:
lock can return nullptr. Either check the return pointer or iterate using a counter and getParentEdgeAt() since this method throws an exception if can not lock the pointer.

Reply:
I think the plugin's code in general is missing functions:
```c++
MKLDNNNodePtr getParentNode(MKLDNNEdgeWeakPtr edge);
MKLDNNNodePtr getChildNode(MKLDNNEdgeWeakPtr edge);
MKLDNNNodePtr getParentEdge(MKLDNNEdgeWeakPtr edge);
MKLDNNNodePtr getChildEdge(MKLDNNEdgeWeakPtr edge);
```

So we won't have to deal with indices for no reason.

Reply:
Let's write them then.

Reply:
Started to implement and than decided to not mix it with the change, since ideally this would required updating a lot of similar code pieces.